extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -210,29 +214,43 @@ export class Bind {
    * Parses and evaluates an expression with a given scope and merges the
    * resulting object into current state.
    * @param {string} expression
-   * @param {!Object} scope
+   * @param {!JsonObject} scope
    * @return {!Promise}
    */
   setStateWithExpression(expression, scope) {
-    this.setStatePromise_ = this.initializePromise_.then(() => {
-      // Allow expression to reference current scope in addition to event scope.
-      Object.assign(scope, this.scope_);
-      return this.ww_('bind.evaluateExpression', [expression, scope]);
-    }).then(returnValue => {
-      const {result, error} = returnValue;
-      if (error) {
-        const userError = user().createError(`${TAG}: AMP.setState() failed `
-            + `with error: ${error.message}`);
-        userError.stack = error.stack;
-        reportError(userError);
-      } else {
-        return this.setState(result);
-      }
-    });
+    this.setStatePromise_ = this.evaluateExpression_(expression, scope)
+        .then(result => this.setState(result));
     return this.setStatePromise_;
   }
 
   /**
+   * Same as setStateWithExpression() except also pushes new history.
+   * Popping the new history stack entry will restore the values of variables
+   * in `expression`.
+   * @param {string} expression
+   * @param {!JsonObject} scope
+   * @return {!Promise}
+   */
+  pushStateWithExpression(expression, scope) {
+    return this.evaluateExpression_(expression, scope).then(result => {
+      // Store the current values of each referenced variable in `expression`
+      // so that we can restore them on history-pop.
+      const oldState = map();
+      Object.keys(result).forEach(variable => {
+        const s = map();
+        const value = this.state_[variable];
+        s[variable] = (value === undefined) ? null : value;
+        deepMerge(oldState, s, MAX_MERGE_DEPTH);
+      });
+
+      const onPop = () => this.setState(oldState);
+      this.history_.push(onPop);
```

Comment:
Oh I see. Don't users have the large "X" button in the top left to return to SRP? I'm not sure we should install a general constraint for a single AMP surface (SRP), especially as we push AMP embeds, AMP as canonical, etc.

Implementation would also need to be complicated to support browser restarts and restoring tabs (storing current stack size in memory won't suffice). 

I'll start a separate discussion issue for this and loop in the PMs.